packages/ui/Button/Button.js

```diff
@@ -167,10 +173,34 @@ const ButtonBase = kind({
 			pressed,
 			selected
 		}, size),
-		icon: ({css, icon, iconComponent: Icon, size}) => {
-			return (typeof icon === 'string' && Icon) ? (
-				<Icon size={size} className={css.icon}>{icon}</Icon>
-			) : icon;
+		icon: ({css, icon, iconComponent, size}) => {
+			if (icon == null || icon === false) return;
+			// If the iconComponent is a string, we exclude the obviously unsupported props
+			// (non-HTML) from the component to simplify the usage of plain HTML elements as the
+			// component. Ex: 'div', 'span', etc.
+			if (typeof iconComponent === 'string') {
+				return (
+					<ComponentOverride
+						component={iconComponent}
+						className={css.icon}
+					>
+						{icon}
+					</ComponentOverride>
+				);
+			} else {
+				// Relay the full compliment of props to the iconComponent. Full Enact components
+				// should be responsible for handling these additional props themselves.
+				// console.log('Rendering %s as a %s icon:', icon, size);
+				return (
+					<ComponentOverride
+						component={iconComponent}
+						size={size}
+						className={css.icon}
+					>
+						{icon}
+					</ComponentOverride>
+				);
+			}
```

Comment:
These returns are very similar, just one prop difference?  What about something like:
```
const props = {
    className: css.icon,
    component: iconComponent,
    size
};

if (typeof iconComponent === 'string') {
    delete props.size;
}

return (
    <ComponentOverride {...props}>
        {icon}
    </ComponentOverride>
);
```

Could also add `children: icon` to those props for a one-line return.

Reply:
Yeah, this is a better approach. Originally, during the development of this, there were more props, so it was more appropriate earlier. I'll update.